### PR TITLE
Transition `clone()` to core API

### LINKED ISSUE #3538: Undesired submodule url
Originally at https://github.com/datalad/datalad/pull/2601

`datalad install -d . -s ../raw sourcedata --reckless`

leads to

```diff
--- /dev/null
+++ b/.gitmodules
@@ -0,0 +1,3 @@
+[submodule "sourcedata"]
+       path = sourcedata
+       url = /home/data/psyinf/scratch/multires3t/bids/../raw
```

### LINKED ISSUE #3958: Unexpected `install/clone` behavior (relpath misinterpretation)
I ran

```
In [1]: import datalad.api as dl                                                             
In [7]: dl.install('~/forrest/collection/phase2', 'dest2')                                   
[INFO   ] Cloning dest2 into '/tmp/~/forrest/collection/phase2' 
Out[7]: <Dataset path=/tmp/~/forrest/collection/phase2>
```

I my case both locations have a dataset, so I expected failure and was surprised by the success. It is mistaking the `~` path for a relative path specification. `clone` has the same issue.

### PR DESCRIPTION
**Pull request by mih**
This is implementing a resultion of https://github.com/datalad/datalad/issues/3957

- [x] RF to make clone(`alt_sources`) parameter obsolete (new internal helper for `get()` and `clone()`)
- [x] Optimize candidate URL handling (only generate once, do proper progress logging, no adhoc)
- [x] Equalize return value behavior with `install()`
- [x] Turn `--reckless` into a mode-style option (backward-compatible)
- [x] Fix #3958 
- [x] Mitigate #3969
- [x] Fix https://github.com/datalad/datalad/issues/3971
- [x] promote `_clone_dataset()` to `clone_dataset()` for use by extensions and other expert Python users
- [x] configure origin of source as a remote https://github.com/datalad/datalad/issues/3926
- [x] go beyond #3926 and discover all traceable local dataset siblings as remotes
- [x] move to `datalad.core.distributed` following https://github.com/datalad/datalad/issues/3192
- [x] add compatibility shim analog to `datalad/interface/run.py`
- [x] investigate https://github.com/datalad/datalad/issues/1146
- [x] investigate https://github.com/datalad/datalad/issues/1786
- [x] investigate https://github.com/datalad/datalad/issues/1956
- [x] beautify relpath submodule url/path. fixes #3538
- [x] investigate #1956
- [x] get testrepos working, so that all tests could run on windows (via #3975)

### CODE REVIEW

**Comment by mih** [comment]
The remaining failure is a curious one. I cannot replicate it locally, and it seems to suggest that `source` and `path` are mistaken for each other?

https://travis-ci.org/datalad/datalad/jobs/632239367?utm_medium=notification&utm_source=github_status
```
======================================================================
4059ERROR: datalad.core.distributed.tests.test_clone.test_clone_dataladri
4060----------------------------------------------------------------------
4061Traceback (most recent call last):
4062  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
4063    self.test(*self.arg)
4064  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/tests/utils.py", line 559, in newfunc
4065    return t(*(arg + (d,)), **kw)
4066  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/tests/utils.py", line 671, in newfunc
4067    return tfunc(*(args + (path, url)), **kwargs)
4068  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/tests/utils.py", line 732, in newfunc
4069    return t(*(arg + (filename,)), **kw)
4070  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/core/distributed/tests/test_clone.py", line 217, in test_clone_dataladri
4071    ds = clone('///ds', path, result_xfm='datasets', return_type='item-or-list')
4072  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/interface/utils.py", line 493, in eval_func
4073    return return_func(generator_func)(*args, **kwargs)
4074  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/interface/utils.py", line 481, in return_func
4075    results = list(results)
4076  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/interface/utils.py", line 468, in generator_func
4077    msg="Command did not complete successfully")
4078datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully [{'message': "could not create work tree dir '/tmp/datalad_temp_test_clone_report_permission_issueju35etyq/protected/datasets-tests.datalad.org': Permission denied", 'path': '/tmp/datalad_temp_test_clone_dataladril6_qiy1k', 'type': 'dataset', 'action': 'install', 'source_url': '///ds', 'status': 'error'}]
4079
```

**Comment by mih** [comment]
Ok, reran that test and it passed. Nice!

**Comment by mih** [comment]
@adswa I have pushed a small update of the `clone` docs (incl. examples taken from `install`). Could you please have a look at the command docs, whether you can spot anything that should be mentioned there (or differently), but isn't. Thx!

**Comment by adswa** [comment]
Looks good to me!

File: datalad/core/distributed/clone.py [hunk 1]
```diff
@@ -0,0 +1,625 @@
...
+    # generate candidate URLs from source argument to overcome a few corner cases
+    # and hopefully be more robust than git clone
+    candidate_sources = [
+        s for src in srcs
+        for s in _get_flexible_source_candidates(src)
+    ]
+
+    # important test! based on this `rmtree` will happen below after failed clone
+    dest_path_existed = dest_path.exists()
+    if dest_path_existed and any(dest_path.iterdir()):
```
Comment (bpoldrack) [inline datalad/core/distributed/clone.py]: Going through this entire PR, this line makes me wonder. As far as I can tell, we never tested destination path to be a directory. If it isn't, then how does that line fail? `AttributeError` (no attribute `iterdir`)? Guess it could simply be included in sanity checks within `Clone.__call__`.

Comment (bpoldrack) [reply datalad/core/distributed/clone.py]: Although, thinking about it. It probably should be tested by `Dataset` or `*Repo` constructors. If path exists, it has to be a directory.

File: datalad/core/distributed/clone.py [hunk 2]
```diff
@@ -0,0 +1,625 @@
...
+    if reckless:
+        lgr.debug(
+            "Instruct annex to hardlink content in %s from local "
+            "sources, if possible (reckless)", ds.path)
+        ds.config.add(
+            'annex.hardlink', 'true', where='local', reload=True)
+
+    # we have just cloned the repo, so it has 'origin', configure any
+    # reachable origin of origins
+    yield from configure_origins(ds, ds, label=None)
```
Comment (bpoldrack) [inline datalad/core/distributed/clone.py]: I feel like that call to `configure_origin` doesn't belong in `_handle_possible_annex_dataset`. As far as I understand what it's for, it's not related to the cloned repo having or not having an annex.
So, move that `yield from` up to line 450 or sth?

Comment (mih) [reply datalad/core/distributed/clone.py]: It only makes sense for annex datasets.

File: datalad/core/distributed/clone.py [hunk 3]
```diff
@@ -0,0 +1,625 @@
...
+            # but since humans might read it, we better confuse them with our
+            # own terms!
+            single_or_plural("dataset sibling", "dataset siblings", len(srs[False]), True),
+            ", ".join(srs[False]),
+            ds.path,
+            srs[False][0] if len(srs[False]) == 1 else "SIBLING",
+        )
+
+
+def configure_origins(cfgds, probds, label):
```
Comment (bpoldrack) [inline datalad/core/distributed/clone.py]: This needs a docstring. Parameters aren't self-explanatory to me.

File: datalad/core/distributed/clone.py (hunk 2 above)
Comment (bpoldrack) [reply datalad/core/distributed/clone.py]: Does it? I get, that availability propagation is the purpose, but at the same time - it actually does care for remotes only and one might as well be interested in other branches than `git-annex`. I could see it being confusing to have those remotes automatically available when there is an annex but not when it's plain git. All that functionality relies on, is plain git mechanism. And for PR workflows like ours, I do find such a functionality useful even with plain git repos. The fact that there's no git-annex branch doesn't determine my interest in other remote branches.
How does it "not make sense" for git repo to connect to known other repos with a common history? One might explicitly not want that, but that might be true for annexes as well.

**Comment by bpoldrack** [comment]
According to #3957, `--message` is still missing.
Other than that and some minor comments above - that looks all reasonable to me.

File: datalad/core/distributed/tests/test_clone.py [hunk 4]
```diff
@@ -427,3 +427,57 @@ def test_installationpath_from_url():
...
+
+# https://github.com/datalad/datalad/issues/3958
+@with_tempfile(mkdir=True)
+@with_tempfile(mkdir=True)
+def test_expanduser(srcpath, destpath):
+    src = Dataset(Path(srcpath) / 'src').create()
+    dest = Dataset(Path(destpath) / 'dest').create()
+
+    with chpwd(destpath), patch.dict('os.environ', {'HOME': srcpath}):
+        res = clone('~/src', 'dest', result_xfm=None, return_type='list',
```
Comment (kyleam) [inline datalad/core/distributed/tests/test_clone.py]: I guess it doesn't matter functionally (because this test [passes on windows][0]), but for consistency with the calls below should this be `op.join('~', 'src')`?

[0]: https://github.com/datalad/datalad/runs/374362978#step:8:25

**Comment by kyleam** [comment]
Hmm, when I pushed [an update][0] with typo fixes and a fix to a `log_progress` call, Travis had [two][1] [runs][2] that failed with

```
======================================================================
ERROR: datalad.core.distributed.tests.test_clone.test_clone_dataladri
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/tests/utils.py", line 559, in newfunc
    return t(*(arg + (d,)), **kw)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/tests/utils.py", line 671, in newfunc
    return tfunc(*(args + (path, url)), **kwargs)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/tests/utils.py", line 732, in newfunc
    return t(*(arg + (filename,)), **kw)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/core/distributed/tests/test_clone.py", line 217, in test_clone_dataladri
    ds = clone('///ds', path, result_xfm='datasets', return_type='item-or-list')
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/interface/utils.py", line 493, in eval_func
    return return_func(generator_func)(*args, **kwargs)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/interface/utils.py", line 481, in return_func
    results = list(results)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/interface/utils.py", line 468, in generator_func
    msg="Command did not complete successfully")
datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully [{'status': 'error', 'path': '/tmp/datalad_temp_test_clone_dataladri9zqubeyj', 'type': 'dataset', 'message': "could not create work tree dir '/tmp/datalad_temp_test_clone_report_permission_issuetplr3kov/protected/datasets-tests.datalad.org': Permission denied", 'action': 'install', 'source_url': '///ds'}]
```

The Travis build for the previous push (d2a25d090) didn't show this failure, but it seems unlikely that the failure is due to the commits I pushed.

[0]: https://github.com/datalad/datalad/compare/d2a25d0908f6bc9f6ed6558c9078e9c60c762083...f6616bdc8cfa2f4959240d2a32098ecf826bf474
[1]: https://travis-ci.org/datalad/datalad/jobs/633378383#L1111
[2]: https://travis-ci.org/datalad/datalad/jobs/633378386#L4059

File: datalad/core/distributed/clone.py (hunk 2 above)
Comment (mih) [reply datalad/core/distributed/clone.py]: It makes sense, if the outcome is desired. But I am not clearly seeing the use case. What is happening is that any local sibling is discovered by retracing the origin chain. One ends up with lots of `origin-?` remotes. For obtaining file content, once doesn't really care what is what and where things actually come from. For merges and such that is not the same. I very much want to know what is what. If we can come up with a proper use case I am not opposed, but without it, I'd say let's keep it limited to annex.

**Comment by mih** [comment]
> The Travis build for the previous push ([d2a25d0](https://github.com/datalad/datalad/commit/d2a25d0908f6bc9f6ed6558c9078e9c60c762083)) didn't show this failure, but it seems unlikely that the failure is due to the commits I pushed.

True, this test fails randomly, and it did before. No idea re cause.

**Comment by yarikoptic** [comment]
FTR - attempt to build python 3.5.6 out from sources on my laptop (debian buster + testing/unstable) fails - gets stuck in tests with bunch of exceptions alike `ssl.SSLError: [SSL: EE_KEY_TOO_SMALL] ee key too small (_ssl.c:2938)` during tests - seems to be some [libssl compatibility issue](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=900158). will look into establishing a container

File: datalad/core/distributed/clone.py (hunk 1 above)
Comment (mih) [reply datalad/core/distributed/clone.py]: In this context `dest_path` is the `.pathobj` property of a `Dataset` instance. I agree that it is out-of-scope for `clone` to test whether it got a "file dataset".

File: datalad/core/distributed/tests/test_clone.py (hunk 4 above)
Comment (mih) [reply datalad/core/distributed/tests/test_clone.py]: True, done.

File: datalad/core/distributed/clone.py (hunk 3 above)
Comment (mih) [reply datalad/core/distributed/clone.py]: True, done.

File: datalad/core/distributed/clone.py (hunk 2 above)
Comment (mih) [reply datalad/core/distributed/clone.py]: I prefer to let this remain an idea until we discover a use case that makes it desirable. Once that happens, it should be easy to shift this functionality in the code to make it happen for GitRepos too, without imposing breakage in other places.

**Comment by mih** [comment]
I have pushed the requested changes. I would be OK to disable the flaky test on travis. We have more than one alternative in the same file (the only unique part of this test is that it tests datalad RIs without having to have the real `///` around. However, there is value in keeping it fail until the reason was discovered (which is unlikely to be in `clone`).

**Comment by yarikoptic** [comment]
FWIW, you could use
```
singularity pull docker://yarikoptic/sid-py3.5.6
```
to pull yourself a container with python3.5 being available... used sid base image and `nd_freeze` to get back in time, installed python3.5, and then uncommented regular apt lines to install recent git-annex-snapshot.  unfortunately didn't reproduce the issue (yet at least)

File: benchmarks/usecases.py [hunk 5]
```diff
@@ -51,48 +51,48 @@ def time_make_studyforrest_mockup(self):
...
         tnt = public.create('tnt', description='image templates')
-        tnt.clone(source=phase1.path, path=opj('src', 'phase1'), reckless=True)
-        tnt.clone(source=structural.path, path=opj('src', 'structural'), reckless=True)
+        tnt.clone(source=phase1.path, path=opj('src', 'phase1'), reckless='auto')
+        tnt.clone(source=structural.path, path=opj('src', 'structural'), reckless='auto')
         aligned = public.create('aligned', description='aligned image data')
-        aligned.clone(source=phase1.path, path=opj('src', 'phase1'), reckless=True)
-        aligned.clone(source=tnt.path, path=opj('src', 'tnt'), reckless=True)
+        aligned.clone(source=phase1.path, path=opj('src', 'phase1'), reckless='auto')
+        aligned.clone(source=tnt.path, path=opj('src', 'tnt'), reckless='auto')
```
[the remark this answers is not in the document]
Comment (yarikoptic) [reply benchmarks/usecases.py]: what about this one -- would it worth with older codebase?

**Comment by mih** [comment]
Given that the time pressure for this PR is off, I have added a bunch of additional issues that should be investigated in this context before a merge.

File: benchmarks/usecases.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mih) [reply benchmarks/usecases.py]: I can revert this change and leave it at `True`, it will work with old and new code.

[the remark this answers is not in the document]
Comment (mih) [reply benchmarks/usecases.py]: Done

**Comment by mih** [comment]
The flaky test s now disabled.

**Comment by kyleam** [comment]
Updates since I last looked at this, in particular 02e2b4c46 (BF: Avoid relpath mangling for submodule url configuration), look good to me.  

> Given that the time pressure for this PR is off, I have added a bunch of additional issues that should be investigated in this context before a merge.

[my 2 cents] Regarding the remaining to-dos, I dislike as a reviewer when additional features and orthogonal fixes continue to be tacked on to opened PRs (once they're out of draft mode and have already had reviewers' eyes on them).

**Comment by mih** [comment]
> [my 2 cents] Regarding the remaining to-dos, I dislike as a reviewer when additional features and orthogonal fixes continue to be tacked on to opened PRs (once they're out of draft mode and have already had reviewers' eyes on them).

You are right -- I apologize. Let me take this PR apart, and keep only the pieces that have been, and continue in another PR. I will keep the disabling of the flaky test, though.

**Comment by mih** [comment]
Or actually, now that you have seen the additions, I will just move the TODOs into a dedicated issue, and we can merge the beast once the tests pass.

**Comment by kyleam** [comment]
> > [my 2 cents] Regarding the remaining to-dos, I dislike as a reviewer when additional features and orthogonal fixes continue to be tacked on to opened PRs (once they're out of draft mode and have already had reviewers' eyes on them).
> 
> You are right -- I apologize.

That's a very nice response, but I really didn't mean it as anything more than "for this one reviewer, it makes things harder."  There's nothing to apologize for of course.  (How dare you keep fixing and improving things! :>)

> Or actually, now that you have seen the additions, I will just move the TODOs into a dedicated issue, and we can merge the beast once the tests pass.

Sounds good, thanks.
